scene/resources/particles_material.cpp

```diff
@@ -261,6 +269,193 @@ void ParticlesMaterial::_update_shader() {
 	code += "}\n";
 	code += "\n";
 
+	//functions to create simplex noise, curl noise in 3D and 4D
+	code += "vec4 permute(vec4 x) {\n";
+	code += "	return mod(((x * 34.0) + 1.0) * x, 289.0);\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "vec4 taylor_inv_sqrt(vec4 r) {\n";
+	code += "	return 1.79284291400159 - 0.85373472095314 * r;\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "\n\n";
+	code += "float permute_float(float x) {\n";
+	code += "	return floor(mod(((x * 34.0) + 1.0) * x, 289.0));\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "float taylor_inv_sqrt_float(float r) {\n";
+	code += "	return 1.79284291400159 - 0.85373472095314 * r;\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "float snoise_3d(vec3 v) {\n";
+	code += "	vec2 C = vec2(1.0 / 6.0, 1.0 / 3.0) ;\n";
```

Comment:
There are some style issues in the generated code, it should follow the same `clang-format` style guide uses in the `.cpp` and `.glsl` files.

But of course since it's within strings it needs to be done manually here. It's mostly good already, but there are issues like space before `;` (lines 295, 298, 324-5), or space before closing parenthesis (line 297, 308-310, 315).
More occurrences in other functions.

I can imagine why those might have been used for readability, but I think consistency is more important here and the readability doesn't suffer too much.